distributed/tests/test_cancelled_state.py

```diff
@@ -396,3 +397,60 @@ def block_execution(event, lock):
     await lock_executing.release()
 
     assert await fut2 == 2
+
+
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_cancelled_resumed_after_flight_with_dependencies(c, s, w2, w3):
+    # See https://github.com/dask/distributed/pull/6327#discussion_r872231090
+    block_get_data_1 = asyncio.Lock()
+    enter_get_data_1 = asyncio.Event()
+    await block_get_data_1.acquire()
+
+    class BlockGetDataWorker(Worker):
+        def __init__(self, *args, get_data_event, get_data_lock, **kwargs):
+            self._get_data_event = get_data_event
+            self._get_data_lock = get_data_lock
+            super().__init__(*args, **kwargs)
+
+        async def get_data(self, comm, *args, **kwargs):
+            self._get_data_event.set()
+            async with self._get_data_lock:
+                return await super().get_data(comm, *args, **kwargs)
+
+    async with await BlockGetDataWorker(
+        s.address,
+        get_data_event=enter_get_data_1,
+        get_data_lock=block_get_data_1,
+        name="w1",
+    ) as w1:
+
+        f1 = c.submit(inc, 1, key="f1", workers=[w1.address])
+        f2 = c.submit(inc, 2, key="f2", workers=[w1.address])
+        f3 = c.submit(sum, [f1, f2], workers=[w1.address])
+
+        await wait(f3)
+        f4 = c.submit(inc, f3, key="f4", workers=[w2.address])
+
+        await enter_get_data_1.wait()
```

Comment:
Post #6371, you can:
1.
```suggestion
        await wait_for_state(f1.key, "flight", w2)
        await wait_for_state(f2.key, "flight", w2)
```
2. get rid of the BlockGetDataWorker subclass
3. initialise the worker with gen_cluster
4. use
```python
    event = asyncio.Event()
    w1.rpc = _LockedCommPool(w1.rpc, write_event=event)
```

Reply:
I actually like my version with the worker subclass better. The `_LockedCommPool` requires much more low level knowledge and is more brittle in my opinion. I think it should only be used if nothing else is possible.